### ISSUE TITLE
dir gives all entries, even invalid ones

### ISSUE THREAD

**Issue by martindurant** [ContinuumIO/intake#235]
The names of entries in a catalog are allowed to be arbitrary strings. Since attribute access in python requires legal identifier names, only those entries with such names should be given by `dir()` (i.e., only those should show up for tab-completion).

**Comment by danielballan** [comment]
On a related subject, I think that __dir__ presents a problem for large Catalogs, as it would cause the whole list of entries to be loaded even the user hits tab. Could it be removed from RemoteCatalog entirely?

**Comment by martindurant** [comment]
If the user does `dir()` (or `list()`) on the cat, then we don't really have a choice but to give the possibilities, right? The python spec does not allow for these to return lazy things. Similarly for `len()`.

**Comment by martindurant** [comment]
Python docs: 
> "Because dir() is supplied primarily as a convenience for use at an interactive prompt, it tries to supply an interesting set of names more than it tries to supply a rigorously or consistently defined set of names, and its detailed behavior may change across releases. For example, metaclass attributes are not in the result list when the argument is a class."

**Comment by danielballan** [comment]
I agree that the documentation gives us some room for flexibility, such as including what we might call "attributes that are actually items accessible as attributes" in the list.

If I have a Catalog with 100000 entries and I'm worried about what will happen when the user hits TAB, what should I do?

**Comment by martindurant** [comment]
it *could* give the first page, right? Could also give a warning at that time to say that the list is truncated. If they say `list()`, however, there would be no choice.

**Comment by danielballan** [comment]
Yes, I think first page (and any other pages and individual entries we happen to have already pulled down?) is a good best-effort behavior here. Fully agree that if you pass the Catalog to `list()` then you deserve what you get. :-D

> Similarly for len().

I suppose we could add length to the `GET /info` response. For example, a Mongo-backed Catalog could implement `__len__` with a special `count` query rather than pulling up the whole collection into memory.
